In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!pip install geopy  
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes 
import folium # map rendering library

print('Libraries imported.')

Solving environment: ...working... done

# All requested packages already installed.

Libraries imported.


Downloading Data from the wikipedia page into a list "toronto_list"

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
toronto_list= pd.read_html(url)[0]

Data successfully downloaded into the list, Lets display the first records of the list and see how does it looks

In [3]:
toronto_list.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

Now, lets transfer the data from the list into a Dataframe.

In [4]:
CanadaData = pd.DataFrame(toronto_list)

Lets see how many records are in the generated Dataframe.

and also see the records.

In [6]:
CanadaData.shape

(288, 3)

In [8]:
CanadaData.head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

Dropping those records in the CanadaData dataframe with Borough value not assigned.

In [7]:
CanadaData = CanadaData[CanadaData.Borough != "Not assigned"]

checking recrds in the dataframe after dropping records with borough not assigned.

In [8]:
CanadaData.head()

Postcode           Borough     Neighbourhood
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights

Lets check how many records are there after dropping ghe records with Borough Not assigned.

In [9]:
CanadaData.shape

(211, 3)

The following command assign the borough value to the Neighbourhood with lable "Not assigned".

In [14]:
CanadaData.loc[CanadaData.Neighbourhood == 'Not assigned', 'Neighbourhood'] = CanadaData.Borough

In [10]:
CanadaData

Postcode           Borough  \
2        M3A        North York   
3        M4A        North York   
4        M5A  Downtown Toronto   
5        M5A  Downtown Toronto   
6        M6A        North York   
7        M6A        North York   
8        M7A      Queen's Park   
10       M9A         Etobicoke   
11       M1B       Scarborough   
12       M1B       Scarborough   
14       M3B        North York   
15       M4B         East York   
16       M4B         East York   
17       M5B  Downtown Toronto   
18       M5B  Downtown Toronto   
19       M6B        North York   
22       M9B         Etobicoke   
23       M9B         Etobicoke   
24       M9B         Etobicoke   
25       M9B         Etobicoke   
26       M9B         Etobicoke   
27       M1C       Scarborough   
28       M1C       Scarborough   
29       M1C       Scarborough   
31       M3C        North York   
32       M3C        North York   
33       M4C         East York   
34       M5C  Downtown Toronto   
35       M6C              York   
38       M9C         Etobicoke   
39       M9C         Etobicoke   
40       M9C         Etobicoke   
41       M9C         Etobicoke   
42       M1E       Scarborough   
43       M1E       Scarborough   
44       M1E       Scarborough   
47       M4E      East Toronto   
48       M5E  Downtown Toronto   
49       M6E              York   
53       M1G       Scarborough   
56       M4G         East York   
57       M5G  Downtown Toronto   
58       M6G  Downtown Toronto   
62       M1H       Scarborough   
63       M2H        North York   
64       M3H        North York   
65       M3H        North York   
66       M3H        North York   
67       M4H         East York   
68       M5H  Downtown Toronto   
69       M5H  Downtown Toronto   
70       M5H  Downtown Toronto   
71       M6H      West Toronto   
72       M6H      West Toronto   
76       M1J       Scarborough   
77       M2J        North York   
78       M2J        North York   
79       M2J        North York   
80       M3J        North York   
81       M3J        North York   
82       M4J         East York   
83       M5J  Downtown Toronto   
84       M5J  Downtown Toronto   
85       M5J  Downtown Toronto   
86       M6J      West Toronto   
87       M6J      West Toronto   
91       M1K       Scarborough   
92       M1K       Scarborough   
93       M1K       Scarborough   
94       M2K        North York   
95       M3K        North York   
96       M3K        North York   
97       M4K      East Toronto   
98       M4K      East Toronto   
99       M5K  Downtown Toronto   
100      M5K  Downtown Toronto   
101      M6K      West Toronto   
102      M6K      West Toronto   
103      M6K      West Toronto   
107      M1L       Scarborough   
108      M1L       Scarborough   
109      M1L       Scarborough   
110      M2L        North York   
111      M2L        North York   
112      M3L        North York   
113      M4L      East Toronto   
114      M4L      East Toronto   
115      M5L  Downtown Toronto   
116      M5L  Downtown Toronto   
117      M6L        North York   
118      M6L        North York   
119      M6L        North York   
122      M9L        North York   
123      M1M       Scarborough   
124      M1M       Scarborough   
125      M1M       Scarborough   
126      M2M        North York   
127      M2M        North York   
128      M3M        North York   
129      M4M      East Toronto   
130      M5M        North York   
131      M5M        North York   
132      M6M              York   
133      M6M              York   
134      M6M              York   
135      M6M              York   
138      M9M        North York   
139      M9M        North York   
140      M1N       Scarborough   
141      M1N       Scarborough   
142      M2N        North York   
143      M3N        North York   
144      M4N   Central Toronto   
145      M5N   Central Toronto   
146      M6N              York   
147      M6N              York   
150      M9N              

The following command group the Borough and club its Neighbourhood into a single record.

In [11]:
CanadaDataGrouped = CanadaData.groupby(['Postcode', 'Borough'], as_index=False, sort=False).agg(','.join)

Lets display the new Dataframe and check the records after groupby.

In [12]:
CanadaDataGrouped.head()

Postcode           Borough                    Neighbourhood
0      M3A        North York                        Parkwoods
1      M4A        North York                 Victoria Village
2      M5A  Downtown Toronto         Harbourfront,Regent Park
3      M6A        North York  Lawrence Heights,Lawrence Manor
4      M7A      Queen's Park                     Not assigned

Using the provided link to the csv file that has the geographical coordinates of each postal code to get the lattitude and 
longitude values for each postal code in Canada.

In [13]:
LatitudeLongitudeData = pd.read_csv("http://cocl.us/Geospatial_data") 

Lets display the records from the Latitude and Longitude Dataframe.

In [14]:
LatitudeLongitudeData.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

The following command renames the Postal Code column in the LatitudeLongitudeData Dataframe to "Postcode" since the column name 
has to the same for us to use the both LatitudeLongitudeData and CanadaDataGrouped dataframes to include the latitude and 
longitude values for the postcodes available for us in the Canada data.

In [15]:
LatitudeLongitudeData.rename(columns={'Postal Code':'Postcode'}, inplace=True)

Lets check the dataframe once after renaming the column.

In [16]:
LatitudeLongitudeData.head()

Postcode   Latitude  Longitude
0      M1B  43.806686 -79.194353
1      M1C  43.784535 -79.160497
2      M1E  43.763573 -79.188711
3      M1G  43.770992 -79.216917
4      M1H  43.773136 -79.239476

Lets include the Latitude and Longitude coordinates in the Canada dataframe.

In [17]:
CanadaDataMerged = pd.merge(CanadaDataGrouped, LatitudeLongitudeData, on='Postcode')

In [18]:
CanadaDataMerged.head()

Postcode           Borough                    Neighbourhood   Latitude  \
0      M3A        North York                        Parkwoods  43.753259   
1      M4A        North York                 Victoria Village  43.725882   
2      M5A  Downtown Toronto         Harbourfront,Regent Park  43.654260   
3      M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4      M7A      Queen's Park                     Not assigned  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

Extracting Data with Totonto in their Borough and saving the selected data into another Dataframe.

In [26]:
toronto_data = CanadaDataMerged[CanadaDataMerged['Borough'].str.contains("Toronto")].reset_index(drop=True)

In [27]:
toronto_data

Postcode           Borough  \
0       M5A  Downtown Toronto   
1       M5B  Downtown Toronto   
2       M5C  Downtown Toronto   
3       M4E      East Toronto   
4       M5E  Downtown Toronto   
5       M5G  Downtown Toronto   
6       M6G  Downtown Toronto   
7       M5H  Downtown Toronto   
8       M6H      West Toronto   
9       M5J  Downtown Toronto   
10      M6J      West Toronto   
11      M4K      East Toronto   
12      M5K  Downtown Toronto   
13      M6K      West Toronto   
14      M4L      East Toronto   
15      M5L  Downtown Toronto   
16      M4M      East Toronto   
17      M4N   Central Toronto   
18      M5N   Central Toronto   
19      M4P   Central Toronto   
20      M5P   Central Toronto   
21      M6P      West Toronto   
22      M4R   Central Toronto   
23      M5R   Central Toronto   
24      M6R      West Toronto   
25      M4S   Central Toronto   
26      M5S  Downtown Toronto   
27      M6S      West Toronto   
28      M4T   Central Toronto   
29      M5T  Downtown Toronto   
30      M4V   Central Toronto   
31      M5V  Downtown Toronto   
32      M4W  Downtown Toronto   
33      M5W  Downtown Toronto   
34      M4X  Downtown Toronto   
35      M5X  Downtown Toronto   
36      M4Y  Downtown Toronto   
37      M7Y      East Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                            Harbourfront,Regent Park  43.654260 -79.360636  
1                             Ryerson,Garden District  43.657162 -79.378937  
2                                      St. James Town  43.651494 -79.375418  
3                                         The Beaches  43.676357 -79.293031  
4                                         Berczy Park  43.644771 -79.373306  
5                                  Central Bay Street  43.657952 -79.387383  
6                                            Christie  43.669542 -79.422564  
7                              Adelaide,King,Richmond  43.650571 -79.384568  
8                         Dovercourt Village,Dufferin  43.669005 -79.442259  
9     Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
10                            Little Portugal,Trinity  43.647927 -79.419750  
11                        The Danforth West,Riverdale  43.679557 -79.352188  
12            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
13         Brockton,Exhibition Place,Parkdale Village  43.636847 -79.428191  
14                      The Beaches West,India Bazaar  43.668999 -79.315572  
15                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
16                                    Studio District  43.659526 -79.340923  
17                                      Lawrence Park  43.728020 -79.388790  
18                                           Roselawn  43.711695 -79.416936  
19                                   Davisville North  43.712751 -79.390197  
20                 Forest Hill North,Forest Hill West  43.696948 -79.411307  
21                       High Park,The Junction South  43.661608 -79.464763  
22                                 North Toronto West  43.715383 -79.405678  
23                  The Annex,North Midtown,Yorkville  43.672710 -79.405678  
24                              Parkdale,Roncesvalles  43.648960 -79.456325  
25                                         Davisville  43.704324 -79.388790  
26                      Harbord,University of Toronto  43.662696 -79.400049  
27                                  Runnymede,Swansea  43.651571 -79.484450  
28                         Moore Park,Summerhill East  43.689574 -79.383160  
29            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
30  Deer Park,Forest Hill SE,Rathnelly,South Hill,...  43.686412 -79.400049  
31  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
32                                           Rosedale  43.679563 -79.377529  
33                    Stn A PO Boxes 25 The Esplanade  43.64643

In [28]:
toronto_data.shape

(38, 5)

we have 38 reords with toronto borough and all those records are stored in toronto_data dataframe. Now the following code generates the geographical coordinates of Toronto Canada using the geolocator.

In [29]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="Canada_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


The foloowing lines of code generates the map of toronto with the neighbourhoods marked with folium.

In [30]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

The following lines of code define the four square credentials.

In [31]:
CLIENT_ID = 'L2YC5V3P20VYR5G54RNBYOGX5KHVPCESQJNCWOKDENXKZOKS' # your Foursquare ID
CLIENT_SECRET = 'MZSNJU4D3JBWXJWPARX0XU1PQG0DI50L3SAGURYJP2HJXPES' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: L2YC5V3P20VYR5G54RNBYOGX5KHVPCESQJNCWOKDENXKZOKS
CLIENT_SECRET:MZSNJU4D3JBWXJWPARX0XU1PQG0DI50L3SAGURYJP2HJXPES


the following code gets the name of the first neighbourhood of the toronto_data dateframe.

In [33]:
toronto_data.loc[0, 'Neighbourhood']

'Harbourfront,Regent Park'

The following lines of code gets the neighbourhood name and its corresponding lattitude and longitude values from the toronto data and save it into the local variables.

In [35]:
neighborhood_latitude = toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Harbourfront,Regent Park are 43.6542599, -79.3606359.


Defining the Four square API call with the Client Id, Client secret, the latitude and longitude coordinates and the rest of the required values.

In [36]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url

'https://api.foursquare.com/v2/venues/explore?&client_id=L2YC5V3P20VYR5G54RNBYOGX5KHVPCESQJNCWOKDENXKZOKS&client_secret=MZSNJU4D3JBWXJWPARX0XU1PQG0DI50L3SAGURYJP2HJXPES&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

The four square API call would look like the above. the next step is to send the get request and see the results which is being done in the next block of codes.

In [37]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d46b37f6e465000380db131'},
 'response': {'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 47,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.653446723052674,
          'lng': -79.3620167174383}],
        'distance': 143,
       

the result json file would look like the above.

In [67]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

The next block of codes clean the json file and store it into a pandas Dataframe nearby_venues

In [68]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories  \
0                                 Roselle Desserts                Bakery   
1                                    Tandem Coffee           Coffee Shop   
2  Toronto Cooper Koo Family Cherry St YMCA Centre  Gym / Fitness Center   
3                              Body Blitz Spa East                   Spa   
4                               Morning Glory Cafe        Breakfast Spot   

         lat        lng  
0  43.653447 -79.362017  
1  43.653559 -79.361809  
2  43.653191 -79.357947  
3  43.654735 -79.359874  
4  43.653947 -79.361149

In [69]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

47 venues were returned by Foursquare.


47 venues has been returned by the four square API. 

Exploring the neighbourhoods in Toronto.

The following block of code get the venues for each neighbourhood in Toronto.

In [70]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [71]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighbourhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )


Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Adelaide,King,Richmond
Dovercourt Village,Dufferin
Harbourfront East,Toronto Islands,Union Station
Little Portugal,Trinity
The Danforth West,Riverdale
Design Exchange,Toronto Dominion Centre
Brockton,Exhibition Place,Parkdale Village
The Beaches West,India Bazaar
Commerce Court,Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North,Forest Hill West
High Park,The Junction South
North Toronto West
The Annex,North Midtown,Yorkville
Parkdale,Roncesvalles
Davisville
Harbord,University of Toronto
Runnymede,Swansea
Moore Park,Summerhill East
Chinatown,Grange Park,Kensington Market
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Rosedale
Stn A PO Boxes 25 The Esplanade
Cabbagetown,St. James Town
First Canadian Place,Underground city


In [72]:
print(toronto_venues.shape)
toronto_venues.head()

(1699, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Harbourfront,Regent Park               43.65426              -79.360636   
1  Harbourfront,Regent Park               43.65426              -79.360636   
2  Harbourfront,Regent Park               43.65426              -79.360636   
3  Harbourfront,Regent Park               43.65426              -79.360636   
4  Harbourfront,Regent Park               43.65426              -79.360636   

                                             Venue  Venue Latitude  \
0                                 Roselle Desserts       43.653447   
1                                    Tandem Coffee       43.653559   
2  Toronto Cooper Koo Family Cherry St YMCA Centre       43.653191   
3                              Body Blitz Spa East       43.654735   
4                               Morning Glory Cafe       43.653947   

   Venue Longitude        Venue Category  
0       -79.362017                Bakery  
1       -79.361809           Coffee Shop  
2       -79.357947  Gym / Fitness Center  
3       -79.359874                   Spa  
4       -79.361149        Breakfast Spot

In [73]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            55   
Brockton,Exhibition Place,Parkdale Village                             23   
Business Reply Mail Processing Centre 969 Eastern                      17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     17   
Cabbagetown,St. James Town                                             47   
Central Bay Street                                                     81   
Chinatown,Grange Park,Kensington Market                               100   
Christie                                                               16   
Church and Wellesley                                                   85   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             34   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     15   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            18   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      4   
Harbord,University of Toronto                                          34   
Harbourfront East,Toronto Islands,Union Station                       100   
Harbourfront,Regent Park                                               47   
High Park,The Junction South                                           23   
Lawrence Park                                                           3   
Little Portugal,Trinity                                                63   
Moore Park,Summerhill East                                              3   
North Toronto West                                                     19   
Parkdale,Roncesvalles                                                  15   
Rosedale                                                                5   
Roselawn                                                                4   
Runnymede,Swansea                                                      35   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        96   
Studio District                                                        41   
The Annex,North Midtown,Yorkville                                      25   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          20   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             55   
Brockton,Exhibition Place,Parkdale Village                              23   
Business Reply Mail Processing Centre 969 Eastern                       17   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      17   
Cabbagetown,St. James Town                                              47   
Central Bay Street                                                      81   
Chinatown,Grange Park,Kensington Market                                100   
Christie                                                                16   
Church and Wellesley                                                    85   
Commerce Court,Victoria Hotel      

In [74]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 239 uniques categories.


With the above block of codes it could be understood that there are about 239 unique category of venies can be found in Toronto Canada from the total of about 1699 venues in the neighbourhood.

Analysing Each Neighbourhood.

In [75]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0             0         0            0                    0                 0   
1             0         0            0                    0                 0   
2             0         0            0                    0                 0   
3             0         0            0                    0                 0   
4             0         0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               1        0                0         0             0   

   Burrito Place  Bus Line  Bus Stop  Butcher  Café  \
0              0         0         0        0     0   
1              0         0         0        0     0   
2              0         0         0        0     0   
3              0         0         0        0     0   
4              0         0         0        0     0   

   Cajun / Creole Restaurant  Camera Store  Caribbean Restaurant

In [76]:
toronto_onehot.shape

(1699, 239)

In [49]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                         Berczy Park     0.000000   
2          Brockton,Exhibition Place,Parkdale Village     0.000000   
3   Business Reply Mail Processing Centre 969 Eastern     0.058824   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
5                          Cabbagetown,St. James Town     0.000000   
6                                  Central Bay Street     0.012346   
7             Chinatown,Grange Park,Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011765   
10                      Commerce Court,Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
14            Design Exchange,Toronto Dominion Centre     0.000000   
15                        Dovercourt Village,Dufferin     0.000000   
16              First Canadian Place,Underground city     0.000000   
17                 Forest Hill North,Forest Hill West     0.000000   
18                      Harbord,University of Toronto     0.029412   
19    Harbourfront East,Toronto Islands,Union Station     0.000000   
20                           Harbourfront,Regent Park     0.000000   
21                       High Park,The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                            Little Portugal,Trinity     0.015873   
24                         Moore Park,Summerhill East     0.000000   
25                                 North Toronto West     0.052632   
26                              Parkdale,Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                  Runnymede,Swansea     0.000000   
30                            Ryerson,Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.024390   
34                  The Annex,North Midtown,Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                      The Beaches West,India Bazaar     0.000000   
37                        The Danforth West,Riverdale     0.023810   

    Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0            0.000000  0.000000            0.000000      0.000000   
1            0.000000  0.000000            0.000000      0.000000   
2            0.000000  0.000000            0.000000      0.000000   
3            0.000000  0.000000            0.000000      0.000000   
4            0.000000  0.058824            0.058824      0.058824   
5            0.000000  0.000000            0.000000      0.000000   
6            0.000000  0.000000            0.000000      0.000000   
7            0.000000  0.000000            0.000000      0.000000   
8            0.000000  0.000000            0.000000      0.000000   
9            0.011765  0.000000            0.000000      0.000000   
10           0.000000  0.000000            0.000000      0.000000   
11           0.000000  0.000000            0.000000      0.000000   
12           0.000000  0.000000            0.000000      0.000000   
13           0.000000  0.000000            0.000000      0.000000   
14           0.000000  0.000000            0.000000      0.000000   
15           0.000000  0.000000            0.000000      0.000000   
16           0.000000  0.000000            0.000000      0.000000   
17           0.000000  0.000000            0.000000      0.000000   

In [78]:
toronto_grouped.shape

(38, 239)

Toronto grouped gives an idea about the different types of venues in each neighbourhood in Toronto.

the following lines of code gives the top 5 venues in each neighbourhood in Toronto.

In [79]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2       Steakhouse  0.04
3              Bar  0.04
4  Thai Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.09
1        Cocktail Bar  0.05
2              Bakery  0.05
3            Beer Bar  0.04
4  Seafood Restaurant  0.04


----Brockton,Exhibition Place,Parkdale Village----
            venue  freq
0  Breakfast Spot  0.09
1            Café  0.09
2     Coffee Shop  0.09
3          Bakery  0.04
4      Restaurant  0.04


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.12
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3       Garden Center  0.06
4              Garden  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Term

defining the most common venues.

In [80]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Getting the top 10 common venues in each neighbourhood.

In [81]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village           Coffee Shop   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1                Bakery          Cocktail Bar              Beer Bar   
2        Breakfast Spot                  Café             Pet Store   
3           Yoga Studio         Auto Workshop            Comic Shop   
4      Airport Terminal        Airport Lounge         Boat or Ferry   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar        Cosmetics Shop                 Hotel   
1            Steakhouse           Cheese Shop                  Café   
2               Stadium         Burrito Place            Restaurant   
3           Pizza Place            Restaurant               Butcher   
4           Coffee Shop                 Plane       Harbor / Marina   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                   Gym   American Restaurant         Breakfast Spot  
1    Seafood Restaurant        Farmers Market                   Park  
2  Caribbean Restaurant          Climbing Gym          Grocery Store  
3         Burrito Place               Brewery             Skate Park  
4              Boutique                   Bar           Airport Gate

Clustering Neighbourhood using K-means clustering

In [82]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [91]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                             Adelaide,King,Richmond   
1               0                                        Berczy Park   
2               0         Brockton,Exhibition Place,Parkdale Village   
3               0  Business Reply Mail Processing Centre 969 Eastern   
4               0  CN Tower,Bathurst Quay,Island airport,Harbourf...   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Café            Steakhouse   
1           Coffee Shop                Bakery          Cocktail Bar   
2           Coffee Shop        Breakfast Spot                  Café   
3    Light Rail Station           Yoga Studio         Auto Workshop   
4       Airport Service      Airport Terminal        Airport Lounge   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0       Thai Restaurant                   Bar        Cosmetics Shop   
1              Beer Bar            Steakhouse           Cheese Shop   
2             Pet Store               Stadium         Burrito Place   
3            Comic Shop           Pizza Place            Restaurant   
4         Boat or Ferry           Coffee Shop                 Plane   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                 Hotel                   Gym   American Restaurant   
1                  Café    Seafood Restaurant        Farmers Market   
2            Restaurant  Caribbean Restaurant          Climbing Gym   
3               Butcher         Burrito Place               Brewery   
4       Harbor / Marina              Boutique                   Bar   

  10th Most Common Venue  
0         Breakfast Spot  
1                   Park  
2          Grocery Store  
3             Skate Park  
4           Airport Gate

Renaming the "Neighborhood" column in Neighborhood_venues_sorted dataframe to "Neighbourhood" since there is a mismatch in spelling with the corresponding column in toronto_merged, this would make it convienient in processing the data further.

In [93]:
neighborhoods_venues_sorted.rename(columns={'Neighborhood':'Neighbourhood'}, inplace=True)

In [94]:
toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postcode           Borough             Neighbourhood   Latitude  Longitude  \
0      M5A  Downtown Toronto  Harbourfront,Regent Park  43.654260 -79.360636   
1      M5B  Downtown Toronto   Ryerson,Garden District  43.657162 -79.378937   
2      M5C  Downtown Toronto            St. James Town  43.651494 -79.375418   
3      M4E      East Toronto               The Beaches  43.676357 -79.293031   
4      M5E  Downtown Toronto               Berczy Park  43.644771 -79.373306   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               0           Coffee Shop                   Pub   
1               0           Coffee Shop        Clothing Store   
2               0           Coffee Shop                  Café   
3               0     Health Food Store                 Trail   
4               0           Coffee Shop                Bakery   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                  Café                  Park                     Bakery   
1        Cosmetics Shop                  Café  Middle Eastern Restaurant   
2            Restaurant    Italian Restaurant                      Hotel   
3                   Pub           Wings Joint           Doner Restaurant   
4          Cocktail Bar              Beer Bar                 Steakhouse   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0        Breakfast Spot            Restaurant    Mexican Restaurant   
1  Fast Food Restaurant   Japanese Restaurant      Ramen Restaurant   
2          Cocktail Bar        Breakfast Spot              Beer Bar   
3                 Diner        Discount Store              Dive Bar   
4           Cheese Shop                  Café    Seafood Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0            Shoe Store         Farmers Market  
1             Bookstore                  Diner  
2        Clothing Store              Gastropub  
3               Dog Run    Dumpling Restaurant  
4        Farmers Market                   Park

Generating a map of the processed Neighbourhood data using folium to visualize and understand the analysis.

In [96]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Now, now the followong lines of code is to  examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster.

Cluster #1

In [97]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3       East Toronto               0     Health Food Store   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0         Grocery Store   
7   Downtown Toronto               0           Coffee Shop   
8       West Toronto               0           Supermarket   
9   Downtown Toronto               0           Coffee Shop   
10      West Toronto               0                   Bar   
11      East Toronto               0      Greek Restaurant   
12  Downtown Toronto               0           Coffee Shop   
13      West Toronto               0           Coffee Shop   
14      East Toronto               0        Sandwich Place   
15  Downtown Toronto               0           Coffee Shop   
16      East Toronto               0                  Café   
19   Central Toronto               0                   Gym   
21      West Toronto               0    Mexican Restaurant   
22   Central Toronto               0        Clothing Store   
23   Central Toronto               0        Sandwich Place   
24      West Toronto               0        Breakfast Spot   
25   Central Toronto               0        Sandwich Place   
26  Downtown Toronto               0                  Café   
27      West Toronto               0           Pizza Place   
29  Downtown Toronto               0                  Café   
30   Central Toronto               0                   Pub   
31  Downtown Toronto               0       Airport Service   
33  Downtown Toronto               0           Coffee Shop   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0           Coffee Shop   
36  Downtown Toronto               0           Coffee Shop   
37      East Toronto               0    Light Rail Station   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                             Pub                  Café                  Park   
1                  Clothing Store        Cosmetics Shop                  Café   
2                            Café            Restaurant    Italian Restaurant   
3                           Trail                   Pub           Wings Joint   
4                          Bakery          Cocktail Bar              Beer Bar   
5                  Ice Cream Shop    Italian Restaurant        Sandwich Place   
6                            Café                  Park                 Diner   
7                            Café            Steakhouse       Thai Restaurant   
8                          Bakery              Pharmacy                  Bank   
9                           Hotel              Aquarium                  Café   
10                    Coffee Shop      Asian Restaurant           Pizza Place   
11                    Coffee Shop        Ice Cream Shop    Italian Restaurant   
12                           Café                 Hotel            Restaurant   
13                 Breakfast Spot                  Café             Pet Store   
14                            Pub         Burrito Place     Food & Drink Shop   
15                           Café                 Hotel   American Restaurant   
16                    Coffee Shop                Bakery    Italian Restaurant   
19                  Grocery Store                  Park        Sandwich Place   
21                           Café                   Bar         Grocery Store   
22                    Coffee Shop   Sporting Goods Shop  Gym / Fitness Center   
23                           Café           Coffee Shop           Pizza Place   
24                      Gift Shop            Restaurant          Dessert Shop   
25                   Dessert Shop      Su

Cluster #2

In [98]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Central Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17           Swim School              Bus Line           Wings Joint   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                 Diner    Falafel Restaurant           Event Space   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
17  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant

Cluster #3

In [100]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
20   Central Toronto               2         Jewelry Store   
32  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20                  Park      Sushi Restaurant                 Trail   
32            Playground                 Trail              Building   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20           Wings Joint              Dive Bar               Dog Run   
32           Wings Joint            Donut Shop        Discount Store   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
20      Doner Restaurant            Donut Shop  Eastern European Restaurant  
32              Dive Bar               Dog Run             Doner Restaurant

Cluster #4

In [99]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               3                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18        Ice Cream Shop          Home Service                  Pool   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18      Doner Restaurant                 Diner        Discount Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18              Dive Bar               Dog Run    Dumpling Restaurant

Cluster #5

In [101]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
28  Central Toronto               4            Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
28            Playground          Tennis Court      Department Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
28           Event Space  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
28  Eastern European Restaurant   Dumpling Restaurant             Donut Shop